demo_apps/messenger_llama2.md

```diff
@@ -0,0 +1,192 @@
+
+![](messenger_api_settings.png)
+
+## Writing Llama 2 Enabled Web App
+
+The Python-based web app we developed uses [LangChain](https://www.langchain.com/), an open source LLM development framework, and [Replicate](https://replicate.com/), a service provider hosting LLM models in the cloud, to receive the user query sent by the webhook, which will be covered in the next section, pass the query to Llama 2, and send the Llama 2 answer back to the webhook. For more information on how to use LangChain or LlamaIndex, another LLM app building framework, and other Llama cloud providers or on-premise deployment options to develop Llama 2 apps, see the [Llama 2 demo apps](https://github.com/facebookresearch/llama-recipes/tree/main/demo_apps).
+
+First, let's create a new conda (or you can use venv if you like) environment and install all the required packages:
+
+```
+conda create -n messenger-llama python=3.8
+conda activate messenger-llama
+pip install langchain replicate flask requests uvicorn gunicorn
+```
+
+Then, create a Python file named [llama_messenger.py](llama_messenger.py) that creates a Llama 2 instance and defines an HTTP method `msgrcvd_page` to:
+
+1. receive the user message forwarded by the webhook;
+2. ask Llama 2 for the answer;
+3. send the answer back to the sender using the Facebook graph API.
+
+```
+import langchain
+from langchain.llms import Replicate
+
+from flask import Flask
+from flask import request
+import os
+import requests
+import json
+
+os.environ["REPLICATE_API_TOKEN"] = "<your replicate api token"    
+llama2_13b_chat = "meta/llama-2-13b-chat:f4e2de70d66816a838a89eeeb621910adffb0dd0baba3976c96980970978018d"
+
+llm = Replicate(
+    model=llama2_13b_chat,
+    model_kwargs={"temperature": 0.01, "top_p": 1, "max_new_tokens":500}
+)
+
+app = Flask(__name__)
+
+@app.route('/msgrcvd_page', methods=['POST', 'GET'])
+def msgrcvd_page():    
+    message = request.args.get('message')
+    sender = request.args.get('sender')
+    recipient = request.args.get('recipient')
+
+    answer = llm(message)
+
+    url = f"https://graph.facebook.com/v18.0/{recipient}/messages"
+    params = {
+        'recipient': '{"id": ' + sender + '}',
+        'message': json.dumps({'text': answer}),
+        'messaging_type': 'RESPONSE',
+        'access_token': '<page_access_token>'
+    }
+    headers = {
+        'Content-Type': 'application/json'
+    }
+    response = requests.post(url, params=params, headers=headers)
+
+    return message + "<p/>" + answer
+```
+
+Replace <page_access_token> with the access token copied in step 5 "Open Messenger's API Settings" of the previous section. Now it's time to modify the webhook to complete the whole app.
+
+## Modifying the Webhook 
+
+Open your glitch.com webhook URL created earlier, and change your `app.js` to simply forward the user message and the user and page ids sent by the Messenger Platform to the Llama 2 enabled web app `llama_messenger.py` described in the previous section:
```

Comment:
What is the purpose of the web hook needed if it just forwards the message? Can the messenger platform not be used to send the message directly to the web app?

Reply:
Ideally the webhook should do what the web app does, or the other way around. But I haven't simplified this yet, and am just using the WhatsApp/Messenger doc to set up a JS based webhook, which then calls the Python web app to invoke Llama. I did mention this in https://github.com/facebookresearch/llama-recipes/blob/main/demo_apps/whatsapp_llama2.md:

Note: It's possible and even recommended to implement a webhook in Python and call the Llama directly inside the webhook, instead of making an HTTP request, as the JavaScript code above does, to a Python app which calls Llama and sends the answer to WhatsApp.

I'll try to improve this in a future PR update?

Reply:
I mean, making the call without the intermediate stage of the webhook will make your system less complex and when its recommended it probably the way to go. You think its a lot of work?

Reply:
Agreed in general but in this case (and the WhatsApp Llama2 tutorial too) the JS based Glitch is the recommended webhook provider - I'd need to port it to Python and test with the webhook configuration in the API setting which would take at least half a day, so will leave it for a future update.